File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/phb.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../game_rules'

module Handlers
  # Player's Handbook rules that don't fit elsewhere: spell slot tables,
  # long rest recovery, and carrying capacity.
  module Phb
    WIZARD_SPELL_SLOTS = {
      5 => { '1' => 4, '2' => 3, '3' => 2 }
    }.freeze

    module_function

    def spell_slots(body)
      klass = body['class']
      level = body['level']

      raise HttpError.new(400, 'class must be a string') unless klass.is_a?(String)
      raise HttpError.new(400, 'level must be an integer') unless level.is_a?(Integer)
      raise HttpError.new(400, "unsupported class: #{klass}") unless klass == 'wizard'
      raise HttpError.new(400, "unsupported level: #{level}") unless WIZARD_SPELL_SLOTS.key?(level)

      [200, { class: klass, level: level, slots: WIZARD_SPELL_SLOTS[level] }]
    end

    def long_rest(body)
      level = body['level']
      hp_current = body['hp_current']
      hp_max = body['hp_max']
      hit_dice_spent = body['hit_dice_spent']
      exhaustion_level = body['exhaustion_level']

      raise HttpError.new(400, 'level must be an integer') unless level.is_a?(Integer)
      raise HttpError.new(400, 'hp_current must be an integer') unless hp_current.is_a?(Integer)
      raise HttpError.new(400, 'hp_max must be an integer') unless hp_max.is_a?(Integer)
      raise HttpError.new(400, 'hit_dice_spent must be an integer') unless hit_dice_spent.is_a?(Integer)
      raise HttpError.new(400, 'exhaustion_level must be an integer') unless exhaustion_level.is_a?(Integer)

      restore_amount = [level / 2, 1].max
      new_hit_dice_spent = [hit_dice_spent - restore_amount, 0].max
      new_exhaustion_level = [exhaustion_level - 1, 0].max

      [200, {
        hp_current: hp_max,
        hit_dice_spent: new_hit_dice_spent,
        exhaustion_level: new_exhaustion_level
      }]
    end

    def equipment_load(body)
      strength = body['strength']
      weight = body['weight']

      raise HttpError.new(400, 'strength must be numeric') unless strength.is_a?(Numeric)
      raise HttpError.new(400, 'weight must be numeric') unless weight.is_a?(Numeric)

      capacity = GameRules.numeric_json(strength * 15)
      encumbered = weight > capacity

      [200, { capacity: capacity, weight: GameRules.numeric_json(weight), encumbered: encumbered }]
    end
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/phb.ts
/** Player's Handbook rules: spell slots, long rests, and carrying capacity. */

import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

// Only wizard level 5 is supported; other class/level combinations are rejected explicitly.
const WIZARD_SPELL_SLOTS: Record<number, Record<string, number>> = {
  5: { '1': 4, '2': 3, '3': 2 },
};

export function spellSlots(body: JsonValue): ApiResult {
  const klass = body.class;
  if (typeof klass !== 'string' || klass.length === 0) {
    return { status: 400, body: { error: 'class must be a non-empty string' } };
  }

  const level = body.level;
  if (!isValidIntInRange(level, 1, 20)) {
    return { status: 400, body: { error: 'level must be an integer from 1 through 20' } };
  }

  if (klass !== 'wizard' || !(level in WIZARD_SPELL_SLOTS)) {
    return { status: 400, body: { error: 'unsupported class/level combination' } };
  }

  return { status: 200, body: { class: klass, level, slots: WIZARD_SPELL_SLOTS[level] } };
}

export function longRest(body: JsonValue): ApiResult {
  const level = body.level;
  if (!isValidIntInRange(level, 1, 20)) {
    return { status: 400, body: { error: 'level must be an integer from 1 through 20' } };
  }

  const hpMax = body.hp_max;
  if (!isValidIntInRange(hpMax, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'hp_max must be a non-negative integer' } };
  }

  const hpCurrent = body.hp_current;
  if (!isValidIntInRange(hpCurrent, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'hp_current must be a non-negative integer' } };
  }

  const hitDiceSpent = body.hit_dice_spent;
  if (!isValidIntInRange(hitDiceSpent, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'hit_dice_spent must be a non-negative integer' } };
  }

  const exhaustionLevel = body.exhaustion_level;
  if (!isValidIntInRange(exhaustionLevel, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'exhaustion_level must be a non-negative integer' } };
  }

  const maxRecoverable = Math.max(1, Math.floor(level / 2));
  const recovered = Math.min(hitDiceSpent, maxRecoverable);

  return {
    status: 200,
    body: {
      hp_current: hpMax,
      hit_dice_spent: hitDiceSpent - recovered,
      exhaustion_level: Math.max(0, exhaustionLevel - 1),
    },
  };
}

export function equipmentLoad(body: JsonValue): ApiResult {
  const strength = body.strength;
  if (!isValidIntInRange(strength, 1, 30)) {
    return { status: 400, body: { error: 'strength must be an integer from 1 through 30' } };
  }

  const weight = body.weight;
  if (typeof weight !== 'number' || !Number.isFinite(weight) || weight < 0) {
    return { status: 400, body: { error: 'weight must be a non-negative number' } };
  }

  const capacity = strength * 15;
  return {
    status: 200,
    body: { capacity, weight, encumbered: weight > capacity },
  };
}
